client/signup/config/flows.js

```diff
@@ -16,32 +17,49 @@ import { addQueryArgs } from 'calypso/lib/url';
 import { generateFlows } from 'calypso/signup/config/flows-pure';
 import stepConfig from './steps';
 
-function getCheckoutUrl( dependencies, localeSlug, flowName ) {
+function constructBackUrlFromPath( path ) {
+	if ( config( 'env' ) !== 'production' ) {
+		const protocol = config( 'protocol' ) ?? 'https';
+		const port = config( 'port' ) ? ':' + config( 'port' ) : '';
+		const hostName = config( 'hostname' );
+
+		return `${ protocol }://${ hostName }${ port }${ path }`;
+	}
+
+	return `https://${ config( 'hostname' ) }${ path }`;
+}
+
+function getCheckoutUrl( dependencies, localeSlug, flowName, destination ) {
 	let checkoutURL = `/checkout/${ dependencies.siteSlug }`;
 
 	// Append the locale slug for the userless checkout page.
 	if ( 'no-site' === dependencies.siteSlug && true === dependencies.allowUnauthenticated ) {
 		checkoutURL += `/${ localeSlug }`;
 	}
 
-	let checkoutBackUrl = `https://${ config( 'hostname' ) }/start/${ flowName }/domain-only`;
-	if ( config( 'env' ) !== 'production' ) {
-		const protocol = config( 'protocol' ) ?? 'https';
-		const port = config( 'port' ) ? ':' + config( 'port' ) : '';
-		const hostName = config( 'hostname' );
-
-		checkoutBackUrl = `${ protocol }://${ hostName }${ port }/start/${ flowName }/domain-only`;
+	let isDomainOnly;
+	// TODO:
+	// the domain only flow has special rule. Ideally they should also be configurable in flows-pure.
+	if ( [ 'domain', 'domain-for-gravatar' ].includes( flowName ) ) {
+		isDomainOnly = 1;
+		destination = `/start/${ flowName }/domain-only`;
 	}
 
+	// checkoutBackUrl is required to be a complete URL, and will be further sanitized within the checkout package.
+	// Due to historical reason, `destination` can be either a path or a complete URL.
+	// Thus, if it is determined as not an URL, we assume it as a path here. We can surely make it more comprehensive,
+	// but the required effort and computation cost might outweigh the gain.
+	const checkoutBackUrl = isURL( destination )
+		? destination
+		: constructBackUrlFromPath( destination );
```

Comment:
I think a quick way to avoid reassigning `destination` ([comment](https://github.com/Automattic/wp-calypso/pull/92168/files#r1660478512)) is to pass it to the  `constructBackUrlFromPath` and move the `isURL` check there. We'd also rename it to `constructBackUrl` in that case. So something like:

```
constructBackUrl( [ 'domain', 'domain-for-gravatar' ].includes( flowName ) ? `/start/${ flowName }/domain-only` : destination )
```

Quick thought. Other ways too.